File: dttools/src/work_queue_example.c
/*
Copyright (C) 2008- The University of Notre Dame
This software is distributed under the GNU General Public License.
See the file COPYING for details.
*/

/*
This program is a very simple example of how to use the Work Queue.
It accepts a list of files on the command line.
Each file is compressed with gzip and returned to the user.
*/

#include "work_queue.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

int main(int argc, char *argv[])
{
	struct work_queue *q;
	struct work_queue_task *t;
	int port = WORK_QUEUE_DEFAULT_PORT;
	int taskid;
	int i;

	if(argc < 2) {
		printf("work_queue_example <executable> <file1> [file2] [file3] ...\n");
		printf("Each file given on the command line will be compressed using a remote worker.\n");
		return 0;
	}

	q = work_queue_create(port);
	if(!q) {
		printf("couldn't listen on port %d: %s\n", port, strerror(errno));
		return 1;
	}

	printf("listening on port %d...\n", work_queue_port(q));

	for(i = 1; i < argc; i++) {

		char infile[256], outfile[256], command[256];

		sprintf(infile, "%s", argv[i]);
		sprintf(outfile, "%s.gz", argv[i]);
		sprintf(command, "./gzip < %s > %s", infile, outfile);

		t = work_queue_task_create(command);
		work_queue_task_specify_file(t, "/usr/bin/gzip", "gzip", WORK_QUEUE_INPUT, WORK_QUEUE_CACHE);
		work_queue_task_specify_file(t, infile, infile, WORK_QUEUE_INPUT, WORK_QUEUE_NOCACHE);
		work_queue_task_specify_file(t, outfile, outfile, WORK_QUEUE_OUTPUT, WORK_QUEUE_NOCACHE);
		taskid = work_queue_submit(q, t);

		printf("submitted task (id# %d): %s\n", taskid, t->command_line);
	}

	printf("waiting for tasks to complete...\n");

	while(!work_queue_empty(q)) {

		t = work_queue_wait(q, 5);
		if(t) {
			printf("task (id# %d) complete: %s (return code %d)\n", t->taskid, t->command_line, t->return_status);
			work_queue_task_delete(t);
		}
	}

	printf("all tasks complete!\n");

	work_queue_delete(q);

	return 0;
}

File: work_queue/src/golang/work_queue_example_go/work_queue_example.go
package main

import (
	"fmt"
	"log"
	"os"

	wq "work_queue"
)

func main() {
	if len(os.Args) < 2 {
		log.Println("work_queue_example <file1> [file2] [file3] ...")
		log.Println(`Each file given on the command line will be compressed 
using a remote worker.`)
		return
	}

	gzipPath := "/bin/gzip"
	if _, err := os.Stat(gzipPath); os.IsNotExist(err) {
		log.Fatal(`gzip was not found. Please modify the gzip_path variable 
accordingly. To determine the location of gzip, from the terminal type: 
which gzip (usual locations are /bin/gzip and /usr/bin/gzip)`)
	}

	q := wq.WorkQueueCreate(9123)
	if q == nil {
		log.Fatal("couldn't listen on port 9123\n")
	}
	log.Println("listening on port 9123...")

	for i := 1; i < len(os.Args); i++ {
		inFile := os.Args[i]
		outFile := fmt.Sprintf("%s.gz", os.Args[i])
		command := fmt.Sprintf("./gzip < %s > %s", inFile, outFile)

		t := wq.WorkQueueTaskCreate(command)
		wq.WorkQueueTaskSpecifyFile(t, gzipPath, "gzip",
			wq.Work_queue_file_type_t(wq.WORKQUEUEINPUT),
			wq.Work_queue_file_flags_t(wq.WORKQUEUECACHE))
		wq.WorkQueueTaskSpecifyFile(t, inFile, inFile,
			wq.Work_queue_file_type_t(wq.WORKQUEUEINPUT),
			wq.Work_queue_file_flags_t(wq.WORKQUEUENOCACHE))
		wq.WorkQueueTaskSpecifyFile(t, outFile, outFile,
			wq.Work_queue_file_type_t(wq.WORKQUEUEINPUT),
			wq.Work_queue_file_flags_t(wq.WORKQUEUENOCACHE))

		taskID := wq.WorkQueueSubmit(q, t)
		log.Printf("submitted task (id# %d)\n", taskID)
	}

	log.Println("waiting for tasks to complete...")
	for wq.WorkQueueEmpty(q) == 0 {
		completeTask := wq.WorkQueueWait(q, 5)
		if completeTask != nil {
			log.Println("a task complete")
			wq.WorkQueueTaskDelete(completeTask)
		}
	}

	log.Println("all tasks complete!")
	wq.WorkQueueDelete(q)
}
